[DOC-3559] Build delegate with immutable image type tutorial (#2878)
harness/developer-hub: tutorials/platform/secure-delegate-default-to-minimal.md

---
title: Build and set up a delegate with an immutable image type
description: This tutorial shows you how to build and set up a delegate with an immutable image type.
sidebar_label: Move from the delegate minimal image
sidebar_position: 6
keywords: [delegate,delegate image]
---

> Harness recommends that you use the Harness Delegate minimal image (*`yy.mm.xxxxx.minimal`*) when you set up Harness Platform for production use. This image has been thoroughly scanned and is free of any high or critical vulnerabilities. Users focused on security tend to prefer this option. However, the minimal delegate image lacks some default binaries, which require you to configure your delegates and install necessary binaries to ensure that Continuous Deployment (CD) steps function properly and remain vulnerability-free from third-party tools. For information on delegate types, go to [Delegate image types](/docs/platform/delegates/delegate-concepts/delegate-image-types).

Reviewer: ```suggestion
Harness recommends that you use the Harness Delegate minimal image (*`yy.mm.xxxxx.minimal`*) when you set up the Harness Platform for production use. This image has been thoroughly scanned and is free of any high or critical vulnerabilities. Users focused on security tend to prefer this option. 

However, the minimal delegate image lacks some binaries that are required for Continuous Deployment (CD) steps to function properly and remain vulnerability-free from third-party tools. Consequently, using the minimal delegate image requires you to configure your delegates and install necessary binaries. For information on delegate types, go to [Delegate image types](/docs/platform/delegates/delegate-concepts/delegate-image-types).
```

The Harness Delegate minimal image (*`yy.mm.xxxxx.minimal`*) is a lighter, more secure version of the default Harness Delegate image. Its main purpose is to provide an enhanced security profile for users, especially those who prioritize their systems' security. The Harness Delegate minimal images includes the following features.

- **Security Scanned:** The image undergoes rigorous scanning processes to ensure that it is devoid of any high-risk or critical vulnerabilities. This makes it an optimal choice for organizations or users who have stringent security requirements.

> - **Limited** Binaries: Unlike the standard delegate, the minimal image does not come packed with all the default binaries. While this contributes to its lightweight nature and security, it also means that users have additional responsibilities. They must manually configure and add any necessary binaries to make their setup functional.

Reviewer: ```suggestion
- **Limited Binaries:** Unlike the standard delegate, the minimal image does not include all of the default binaries. While this contributes to its lightweight nature and security, it also means that users have additional responsibilities. They must manually configure and add any necessary binaries to make their setup functional.
```

- **User Responsibilities:** Because the minimal delegate image is devoid of the default binaries, users are in charge of tailoring it to their needs. This includes installing specific binaries essential for their CD steps. This level of control also allows users to maintain an updated environment. By installing the latest versions of necessary binaries, they can ensure that the delegate remains free from potential vulnerabilities found in outdated third-party tools.

- **Preferred by Security-Conscious Users:** Due to its clean security slate, many users who prioritize system security gravitate towards the minimal delegate image. By starting with a minimal setup and adding only what is necessary, they can maintain a tighter control over the software and tools present, thus minimizing potential security risks.

> This tutorial shows you how to build and set up a delegate with  an immutable image type (*`yy.mm.xxxxx.minimal`*) for CD.

Reviewer: ```suggestion
This tutorial shows you how to build and set up a minimal delegate image (*`yy.mm.xxxxx.minimal`*) for Harness CD.
```

## Prepare the delegate

1. Prepare your Dockerfile. For more information and instructions, go to [Build custom delegate images with third-party tools](/docs/platform/delegates/install-delegates/build-custom-delegate-images-with-third-party-tools/).
2. Set up Docker locally to be able to build in multiple architectures. For more information, go to [How to build x86 Docker images on an M1 Mac](https://blog.jaimyn.dev/how-to-build-multi-architecture-docker-images-on-an-m1-mac/).
3. Build and push
   
   ```
   docker buildx build --platform linux/amd64 --output type=docker .
   docker image ls
   docker tag 34594c2f0882 harness/delegate:yy.mm.xxxxx.harness.custom.x86.v4
   docker push harness/delegate:yy.mm.xxxxx.harness.custom.x86.v4
   ```

> 4. Update your delegate image in the manifest with the tag from the image you pushed in previous step.

Reviewer: ```suggestion
4. In the delegate image manifest, update the delegate image with the tag from the image you pushed in the previous step.
```

Reviewer: this needs an example, imo. I don't see a manifest with the tag updated, or even a snippet. So how does the user know where to do it?

Author: @michaelcretzman nice catch. Thank you. Worked with @pratmit to add an xref here.

       - secretRef:  
           name: custom-del-account-token  
       env:  
       - name: JAVA_OPTS  
         value: "-Xms64M"  
       - name: ACCOUNT_ID  
         value:   
       - name: MANAGER_HOST_AND_PORT  
         value: https://app.harness.io/gratis  
       - name: DEPLOY_MODE  
         value: KUBERNETES  
       - name: DELEGATE_NAME  
         value: custom-del  
       - name: DELEGATE_TYPE  
         value: "KUBERNETES"  
       - name: DELEGATE_NAMESPACE  
         valueFrom:  
           fieldRef:  
             fieldPath: metadata.namespace  
       - name: INIT_SCRIPT  
         value: ""  
       - name: DELEGATE_DESCRIPTION  
         value: ""  
       - name: DELEGATE_TAGS  
         value: ""  
       - name: NEXT_GEN  
         value: "true"  
       - name: CLIENT_TOOLS_DOWNLOAD_DISABLED  
         value: "true"  
       - name: LOG_STREAMING_SERVICE_URL  
         value: "https://app.harness.io/gratis/log-service/"  
   
---  
   
apiVersion: v1  
kind: Service  
metadata:  
 name: delegate-service  
 namespace: harness-delegate-ng  
spec:  
 type: ClusterIP  
 selector:  
   harness.io/name: custom-del  
 ports:  
   - port: 8080  
   
---  

>    

Reviewer 2: Everything below this line is not needed if upgrader job is removed (original yaml had it commented out)
